src/services/feed_manager.py

```
@@ -65,53 +67,55 @@
-            with open(self.invalid_file, 'w', encoding='utf-8') as f:
-                f.write(xmlstr)
-            
-            logging.info(f"Saved {len(invalid_feeds_by_genre)} invalid feeds to {self.invalid_file}")
 
+    async def _save_invalid_feeds(self, invalid_feeds_by_genre: List[Feed]) -> None:
+        """Save invalid feeds to a separate OPML file."""
+        root = create_opml_tree(
+            invalid_feeds_by_genre,
+            title="Invalid RSS Feeds - Grouped by Original Category",
+        )
+        xmlstr = minidom.parseString(ET.tostring(root)).toprettyxml(indent="  ")
```

Ruff (0.8.0)

Using `xml` to parse untrusted data is known to be vulnerable to XML attacks; use `defusedxml` equivalents

(S318)